diag_manager/diag_manager.F90

```diff
@@ -4496,70 +4493,66 @@ SUBROUTINE diag_field_attribute_init(diag_field_id, name, type, cval, ival, rval
+    INTEGER,          INTENT(in) :: diag_field_id !< ID number for field to add attribute to
+    CHARACTER(len=*), INTENT(in) :: att_name      !< new attribute name
+    class(*),         INTENT(in) :: att_value     !< new attribute value
 
     if (use_modern_diag) then
-      call fms_diag_object%fms_diag_field_add_attribute(diag_field_id, att_name, (/att_value /))
-    else
-      CALL diag_field_add_attribute_r1d(diag_field_id, att_name, (/ att_value /))
-    endif
-  END SUBROUTINE diag_field_add_attribute_scalar_r
-
-  !> @brief Add a scalar integer attribute to the diag field corresponding to a given id
-  SUBROUTINE diag_field_add_attribute_scalar_i(diag_field_id, att_name, att_value)
-    INTEGER, INTENT(in) :: diag_field_id !< ID number for field to add attribute to
-    CHARACTER(len=*), INTENT(in) :: att_name !< new attribute name
-    INTEGER, INTENT(in) :: att_value !< new attribute value
-
-    if (use_modern_diag) then
-      call fms_diag_object%fms_diag_field_add_attribute(diag_field_id, att_name, (/att_value /))
-    else
-      CALL diag_field_add_attribute_i1d(diag_field_id, att_name, (/ att_value /))
-    endif
-  END SUBROUTINE diag_field_add_attribute_scalar_i
-
-  !> @brief Add a scalar character attribute to the diag field corresponding to a given id
-  SUBROUTINE diag_field_add_attribute_scalar_c(diag_field_id, att_name, att_value)
-    INTEGER, INTENT(in) :: diag_field_id !< ID number for field to add attribute to
-    CHARACTER(len=*), INTENT(in) :: att_name !< new attribute name
-    CHARACTER(len=*), INTENT(in) :: att_value !< new attribute value
-
-    if (use_modern_diag) then
-      call fms_diag_object%fms_diag_field_add_attribute(diag_field_id, att_name, (/att_value /))
+      select type(att_value)
+      type is (real(kind=r4_kind))
+         call fms_diag_object%fms_diag_field_add_attribute(diag_field_id, att_name, (/att_value /))
+      type is (real(kind=r8_kind))
+         call fms_diag_object%fms_diag_field_add_attribute(diag_field_id, att_name, (/att_value /))
+      type is (integer(kind=i4_kind))
+         call fms_diag_object%fms_diag_field_add_attribute(diag_field_id, att_name, (/att_value /))
+      type is (character(len=*))
+         call fms_diag_object%fms_diag_field_add_attribute(diag_field_id, att_name, (/att_value /))
+      class default
+         call mpp_error(FATAL, "Diag_field_add_attribute 0d:: unsupported type. The acceptable types "//&
+                            "are float, double, integer, and string")
+      end select
     else
-      CALL diag_field_attribute_init(diag_field_id, att_name, NF90_CHAR, cval=att_value)
+      select type(att_value)
+      type is (real(kind=r4_kind))
+         CALL diag_field_attribute_init(diag_field_id, att_name, NF90_FLOAT, rval=real((/att_value/)))
+      type is (real(kind=r8_kind))
+         CALL diag_field_attribute_init(diag_field_id, att_name, NF90_FLOAT, rval=real((/att_value/)))
```

Comment:
If a r8 value is coming in, shouldn't this be an [`NF90_DOUBLE`](https://docs.unidata.ucar.edu/netcdf-fortran/current/f90-variables.html#f90-language-types-corresponding-to-netcdf-external-data-types) in both?

Reply:
In both routines, not both lines, that is.